Feature/event modeling diagram

lgazo opened the pull request:
## :bookmark_tabs: Summary

This PR implements [Event Modeling](https://eventmodeling.org/) Diagram. 

> Event Modeling is a method of describing systems using an example of how information has changed within them over time.

There was an attempt within #5860 ticket, but I have chosen different approach where writing the text description must be rapid and compact in the first place.

Resolves #5860

## :straight_ruler: Design Decisions

EM Diagram uses a specific layout that is a combination of timeline / sequence diagram, swimlanes and limited amount of Entity types. Relations among Entities are inferred as much as possible so a Designer focuses primarily on capturing the requirements.

I have studied and looked for existing layout algorithms (e.g. ELK, GraphViz,...) but they were not sufficient for this case so far.

The grammar uses Langium. Initially the grammar was developed separately and then brought to Mermaid, where after a few iteration it reached its completeness. The grammar contains more elements than what is currently covered by the renderer.

The renderer and the underlying DB uses Decider pattern to build required state data object out of the AST. In the first iteration it uses the provided D3 renderer but other types of outputs might be possible in the future. The properties of the final diagram are hardcoded at the moment. Colours follow the convention introduced by a standard EM diagram.

No additional external libraries were used.

I assume that further improvements to the implementation might be possible as I have just initial familiarity with Mermaid project and its available components.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

lgazo committed:
feat(eventmodeling): first version of Event Modeling DSL

This initial version contain working parser for the Event Modeling text DSL.
It has also trivial renderer that lays out entities in three fixed
swimlanes. There are no edges yet. It is to prove the whole lifecycle and
get familiar with all the places when implementing new diagram type.

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

lgazo committed:
feat(eventmodeling): box layout polished
* nice box colors
* compressing space between boxes when it makes "visual" sense
* adding types everywhere
* remove copied commented code

lgazo committed:
Merge remote-tracking branch 'upstream/develop' into develop

github-actions[bot] committed:
chore: update E2E timings

github-actions[bot] committed:
chore: update E2E timings

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
Merge remote-tracking branch 'upstream/develop' into develop

lgazo committed:
Merge branch 'develop' of github.com:lgazo/mermaid into develop

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
feat(eventmodeling): relations drawn among boxes
* infer correct source from a box not in the same swimlane
* draw straight lines with arrows

lgazo committed:
chore(eventmodeling): using logger + remove border from boxes

lgazo committed:
Merge remote-tracking branch 'upstream/develop' into develop

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
refactor(eventmodeling): namespacing the event modeling language
* changed also the source frame identifier from >t to >tf to be consistent

lgazo committed:
fix(eventmodeling): reverting wiped out pnpm-lock
* now the build compiles everything correctly
* fixed problem with cytoscape and architect plugin

lgazo committed:
feat(eventmodeling): nicer box colours with a border

lgazo committed:
fix(eventmodeling): missing tests for diagram parser

lgazo committed:
chore(eventmodeling): removing obsolete commented code + aligning on types

lgazo committed:
feat(eventmodeling): swimlane names and dynamic size + data rendered inside boxes
* types moved to `types.ts`
* another round of renderer code cleanup
* swimlanes rendering considering the dimensions of boxes inside
* boxes rendered with corresponding inline or block data entities respecting rendering size

lgazo committed:
feat(eventmodeling): multiple source frames relation rendering
* support to reset the relation inference with `rf` (EmResetFrame) type

lgazo committed:
feat(eventmodeling): dynamic swimlane position and size
* swimlane adapts to the biggest box it contains
* also swimlanes are correctly vertically positioned

lgazo committed:
feat(eventmodeling): render swimlanes for namespaced entities
* base swimlanes split by defined namespaces
* you can have up to 99 swimlanes in a base group (UI/A, C/RM, Event)

lgazo committed:
chore(eventmodeling): improved code coverage of diagrams + refactoring of the data logic to `db`
* moved the data manipulation logic and data model preparation to `db` file
* added test spec for basic parsing in the diagram, parsing tests are covered in `parser` package, no need to duplicate
* rendering coverage did not work as expected, keeping commented

lgazo committed:
fix: last swimlane y-position calculated correctly

lgazo committed:
feat(eventmodeling): documentation

lgazo committed:
fix(eventmodeling): cosmetical issues in the documentation

schicks commented:
Thanks for doing this! I'm on a phone and haven't had a chance to review thoroughly, just looked over the documentation. 


I was surprised to see that time frames were explicit. I am concerned that will make the diagrams hard to edit, since adding something or moving something early in the flow will require lots of changes throughout the definition. What do explicit timeframes get us?

lgazo commented:
@schicks  The approach I took is based on the fact that the EM diagram is basically a timeline with quite well defined transitions / patterns most of the time. It is fairly easy to reshuffle Time Frames because you just move, append, insert or change lines. The order of lines defines the flow, not the numbers. You can put any number you want as long as it is unique. Numbers are useful only for explicit relations. By default the relations should be inferred so you are quicker to design and with less definitions. You have the option to break the inference or define relations manually when needed.

lgazo committed:
fix(eventmodeling): improved coverage of the `db` file

knsv commented:
Thank you for contributing your proposal for an Event Modeling diagram syntax to Mermaid! Your work clearly shows thoughtful consideration of Event Modeling concepts and provides a solid foundation.

While your proposal effectively captures many detailed aspects of Event Modeling, it diverges somewhat from the existing Mermaid diagram syntax conventions, potentially increasing complexity for users who already understand Mermaid diagrams. Ensuring consistency with the Mermaid ecosystem is crucial, as it helps users intuitively grasp and remember new diagram types.

Here are some suggestions for aligning your syntax more closely with Mermaid:

1. **Entity Definitions:**
   Consider using simpler and more concise tokens aligned with existing Mermaid conventions:

   ```
   eventmodeling

   screen CartScreen
   command AddItem
   event ItemAdded
   ```

2. **Data Blocks:**
   Mermaid typically favors concise syntax. Simplify inline data:

```
   command AddItem {description: string, price: number}
   event ItemAdded {description: string, price: number}
```

3. **Flow and Relations:**
   Define relationships explicitly, similar to Mermaid flowcharts:

   ```
   screen CartScreen --> command AddItem --> event ItemAdded
   ```

4. **Namespaces and lanes:**
   Utilize existing Mermaid like lane syntax clearly:

   ```
   eventmodeling

   %% Using indentations like for mindmaps/kanban diagrams
   
   lane Inventory
       event InventoryChanged

   lane External
       event External.InventoryUpdated

	%% or similar to subgraphs 
	lane 
      ...
    end

    // or using brackets

	External {
	  ...
	}
    
   ```

5. **Consistency with Existing Diagrams:**
   Draw inspiration from Mermaid sequence or state diagrams:

   ```
   sequenceDiagram
       CartScreen->>AddItem: trigger
       AddItem->>ItemAdded: produce
   ```

A good next step might be to provide a revised, syntax example demonstrating these alignment suggestions. This will facilitate further discussion and help ensure your contribution is both powerful and intuitive for the broader Mermaid user community.

We greatly appreciate your effort and creativity, and we're excited to collaborate further to integrate Event Modeling effectively into Mermaid.

lgazo commented:
Thank you @knsv for your time taking to review the PR.

First of all I would like to highlight the crucial consideration behind the grammar - that is to provide compact and distraction-less designing using the concepts of EM diagram. Central part of the EM is the timeline, similar to UML Sequence diagram, rotated by 90-degrees, going from left to right. However the significant difference is the repetition of patterns for the key flow in the diagram, which in turn, if considered can save the amount of explicit relationship definitions. Otherwise maybe a more generic type of diagram like State diagram, if extended with the swimlanes and some layouting options, could potentially cover the basic requirements.

Although I am fully aware that the nodes and relations could be defined explicitly like in any other UML diagram, I took different approach for the above reasons. The grammar itself is extremely compact for the timeline part. But is more verbose for the additional definitions, like for Data or Notes or GWTs (that are not yet part of the rendering).

In order to address some of the concerns, I would propose to extend the grammar in second iteration with the following.

1. Compact mode vs Relaxed mode

The grammar can accept both types of tokens, e.g. `tf` and `timeframe`, `scn` and `screen`, etc.

```
eventmodeling

timeframe 01 screen CartScreen
timeframe 02 cmd AddItem { productId: 7 }
timeframe 03 event ItemAdded { productId: 7 }
```

2. Data Blocks

I would need explanation for the comment for this one, as the Inline Data format looks exactly like that. Additionally you can specify a hint in backticks, for what is the content type, but that is optional.

3. Flows and Relations

I see that `>f` is out of the convention for the definition of multiple incoming relations. I'm happy to change it to `->>` which is definitely more in line with the Sequence diagram line types and in the future even extend it to other types of lines.

For the explicit relations, actually this would be required for more complex diagrams sooner or later. But the relationships are among Time frames instead of Entity types. From my point of view it is the addition to the existing grammar. I'd happy to reuse the format, if it would be possible to define in Langium. Something like:

```
timeframe 01 --> timeframe 02
timeframe 04 --> timeframe 02
```

or for Compact mode

```
tf 01 --> tf 02
tf 04 --> tf 02
```

4. Namespaces and lanes

The order of Time frames in the current syntax is important. Grouping them to swimlanes would break the sequence. Of course that applies only in case the current format can stay. 

However, additional configuration options for each swimlane might be required for the future. One of the examples is specific label, icon or additional description. Then I would extend the grammar with a swimlane definition, eg.

```
swimlane Inventory
    label "My Inventory"
    icon "inventory.svg"
```

5. Complex example

```
eventmodeling

timeframe 01 screen CartScreen
// compact and relaxed syntax together
timeframe 02 cmd AddItem { productId: 7 }
timeframe 03 event ItemAdded { productId: 7 }
timeframe 07 event AccountingItemAdded

timeframe 04 readmodel CartItems ->> 03 ->> 07
timeframe 05 rmo Products [[Products05]]

timeframe 08 screen CartScreen
timeframe 09 cmd AddItem { productId: 10 }

// explicit relation
timeframe 06 --> timeframe 01

data Products05 {
  [
    product2,
    product7
  ]
}
```

I can also imagine that the timeline part goes into a separate section, which on the other hand requires more characters:

```
eventmodeling

timeline
  timeframe 01 screen CartScreen
  // compact and relaxed syntax together
  timeframe 02 cmd AddItem { productId: 7 }
  timeframe 03 event ItemAdded { productId: 7 }
  timeframe 07 event AccountingItemAdded

  timeframe 04 readmodel CartItems ->> 03 ->> 07
  timeframe 05 rmo Products [[Products05]]

  timeframe 08 screen CartScreen
  timeframe 09 cmd AddItem { productId: 10 }

// explicit relation
timeframe 06 --> timeframe 01

data Products05 {
  [
    product2,
    product7
  ]
}
```


Summary

I'm a long-term user of different text-to-diagram tools, but the fact is that none of the existing representations fits the EM case (in my opinion). Either it is too verbose or the automated layouting fails due to the specifics of EM diagram. I decided to contribute here only after seeing gitgraph diagram which somewhat reminded me principles of EM.

I would like to find a balance between this design considerations and the conventions from the main diagram types in Mermaid. I would also understand, if this type of grammar is not suitable for Mermaid and it has to live a life of its own elsewhere. The approach I chose needs market validation anyway.

ashishjain0512 commented:
@lgazo Thanks for putting together this PR. Here are my quick notes following the thread above:

1. I like the idea of supporting both compact and relaxed modes.
2. @knsv We need to provide more specific examples where the inline format is not met.
3. For the relations, adopting the existing arrow syntax for relationships `-->` opens up quite a bit on re-using edge types, markers/arrowheads.
4. As @knsv pointed out, Mermaid already has some syntax to support grouping like the subgraphs in flowcharts or indentation-based grouping in mindmaps, or grouping of messages in sequence via a box. Utilizing similar grouping syntax means one less special case, and in the future, when we develop common layout algorithms, it will be easier to apply it to Event Modelling diagrams as well.

Also, for the metadata attributes of any entity/namespace, we could use the new `@` syntax (taking inspiration from flowcharts)
```
flowchart TD
    A@{ icon: "fa:user", form: "square", label: "User Icon", pos: "t", h: 60 }
```

Overall, I think the Event Modelling diagram will be a great fit for Mermaid. 
We try to keep the syntax simple for users to learn and use, and it is important to have this streamlined before we release the new diagram type.

lgazo commented:
Hi @ashishjain0512,
Thank you for your response. Let's sync on the way how to progress further. As I am fairly new to Mermaid code base, I'm not sure if I can make relevant design decisions. It feels that supporting many options provided by the flowchart would require a lot of guidance from someone more experienced. I would suggest to continue ideally in iterations, mainly avoiding the code to be stale for a long while. Additionally, I'm using Langium, but I've seen flowchart uses a different approach. The question is how to combine it.
I'm looking forward for suggestions how to go forward. Feel free to contact me also on Discord, I've joined the project.

diegosasw commented:
Have you checked event modeling in plantuml for inspiration?
https://github.com/chilit-nl/plantuml-event-modeling

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
fix(eventmodeling): compact and relaxed notation + aligned reference token

Support for both compact and relaxed notation implemented. It is interchangeable.
Time Frame reference token uses Sequence diagram convention as suggested in PR.
Documentation updated as well.

lgazo committed:
fix(eventmodeling): visual fixes for data blocks
* correct coloring for relaxed notation
* name of the entity centered by default
* code block used for data blocks
* unified content creation across data inline and data block
* tinkering with visual style

lgazo commented:
> You can not go from Event into UI/Processor; that is invalid, you must define a Read Model. Could your tool enforce that?

@yordis Well, if you refer to the second diagram, it is a translation pattern used by Martin Dilger in his book.

Langium supports custom validations and pattern checks might be feasible to implement. However any unresolved error might block rendering the diagram. Imagine you want to render first draft which contains only screens and events. I can see the value of validators for LSP in an IDE, but that's my intent to cover in a [separate project which contains just the DSL](https://github.com/lgazo/event-modeling-dsl) with potential additional tooling (e.g. interoperable EM data format, other output formats,...).

spearna commented:
@lgazo  I agree with your general stance. It's nice to have a tool validate the flow of the primitives.  However, since Adam and Martin provide two options on the translation pattern flow (one with a read model, one without, respectively) this may need to be done manually until more smarts can be integrated. 

Another question for @lgazo. Since Adam and Martin recommend copying/pasting primitives that are referenced in the flow in order to keep the flow/arrows moving left-to-right, how would that be accomplished in this feature?

lgazo commented:
@spearna that's the key feature I had in mind supported by the timeline. You can see a more [complex example here](https://github.com/lgazo/mermaid-eventmodeling-examples/blob/main/agent_workflow_from_discord.mmd). You can repeat events and read models and additionally you have an option to tie different data to it.

yordis commented:
@dilgerma I care a lot about alignment and precision. Every new interpretation or approach adds branches we all need to analyze and reconcile, which can slow us down. Opinions for the sake opinions, disagreement on silly things. I have been there already, since I have trained at least 20 engineers by now.

If we can stick to fewer, clearly defined concepts, without losing essential context, we’ll make things simpler and more consistent for everyone.

That is something I appreciated when I met Adam back in ~2020 and the reason I adopted it.

And to make it clear, to me, whatever Adam's says is what count, I am not here to tell how it is done, I have done it for such long time that I did not realize some people were doing it differently.

pbrolin47 reviewed (changes requested):
[sisyphus-bot]

Hi @lgazo — thank you so much for this contribution! Event Modeling is a genuinely interesting methodology and the custom layout engine you've built (the Decider/Evolver pattern applied to positioning!) is a creative and well-structured approach. There's real craftsmanship here. Let's get it across the finish line.

I've found a few things that need attention before merge — most importantly a security issue. Here's the full breakdown:

---

## What's working well 🎉

🎉 **Langium grammar is clean and idiomatic** (`event-modeling.langium`) — good use of fragment rules, proper hidden terminals, and the grammar reads clearly. QualifiedName support for namespacing is a nice touch.

🎉 **The Decider/Evolver architecture in `db.ts`** is elegant and testable. Using CQRS-style patterns to separate layout commands from state evolution makes the layout logic much easier to reason about and extend. This is a noticeably better design than a monolithic `draw()` function.

🎉 **Parser tests are solid** (`packages/parser/tests/eventmodeling.test.ts`) — good coverage of multiple frame types, qualified names, data blocks, and multi-source relations. These are realistic, well-chosen scenarios.

---

## Things to address

### 🔴 [blocking] — XSS: unsanitized user content reaches D3 `.html()` via `foreignObject`

`renderer.ts:48–57` uses `foreignObject` + `xhtml:div` + `.html(box.text)`. D3's `.html()` is a direct wrapper around `element.innerHTML` and fires immediately on a live DOM node — **before** the final DOMPurify pass on the serialized SVG runs.

`box.text` is assembled in `db.ts calculateTextProps()` from user-provided values without any sanitization:

```typescript
// db.ts ~line 260
const name = extractName(frame.entityIdentifier); // user text
let content = `<b>${name}</b>`;                   // raw interpolation

// ...dataInlineValue path:
toHtml = frame.dataInlineValue;                   // raw grammar token
toHtml = toHtml.substring(toHtml.indexOf('{') + 1); // stripped, not sanitized

// ...dataBlockValue path:
toHtml = dataEntity.dataBlockValue;               // raw user block
toHtml = toHtml.replaceAll('\n', '<br/>');
toHtml = toHtml.replaceAll(' ', '&nbsp;');        // ← only spaces encoded; < > & are still raw
```

A diagram like:
```
tf 02 evt Foo { <img src=x onerror=alert(document.cookie)> }
```
will fire the `onerror` handler during rendering, before DOMPurify sees the output. The `dataBlockValue` path has the same issue — `replaceAll(' ', '&nbsp;')` gives a false sense of safety but leaves `<`, `>`, and `&` unencoded.

There's a related structural issue: `wrapLabel()` is called on the fully-assembled HTML string (`db.ts` end of `calculateTextProps`). `wrapLabel()` splits on spaces and doesn't understand HTML — it can split inside attribute values, corrupting tag structure in ways that create additional injection vectors.

**Fix:** Follow the pattern used by virtually every other Mermaid diagram:
1. Sanitize user strings at the **input** side in `calculateTextProps()`:
   ```typescript
   import { sanitizeText } from '../../diagrams/common/common.js';
   const name = sanitizeText(extractName(frame.entityIdentifier), getConfig());
   // ...
   toHtml = sanitizeText(toHtml, getConfig());
   ```
2. Call `wrapLabel()` on plain text **before** wrapping in HTML tags.
3. Preferred long-term: replace the `foreignObject`/`.html()` approach with SVG `<text>`/`<tspan>` elements — this removes the HTML injection surface entirely and eliminates the `foreignObject` dependency.

See `diagrams/common/common.ts` for the `sanitizeText()` utility and how `kanban`, `block`, and `class` diagrams use it for label rendering.

---

### 🔴 [blocking] — No Cypress E2E visual regression tests

There are no Cypress test files for this diagram (nothing in `cypress/integration/rendering/eventmodeling*`). Per the project's testing requirements, **any change that affects rendered SVG output requires E2E visual regression tests** using `imgSnapshotTest()`.

Please add a test file at `cypress/integration/rendering/eventmodeling.spec.js` covering at least the three main scenarios: state view, state change, and translation. A minimal example:

```javascript
import { imgSnapshotTest } from '../../helpers/util.ts';
describe('Event Modeling Diagram', () => {
  it('renders a state view pattern', () => {
    imgSnapshotTest(`
      eventmodeling
      tf 01 scn CartScreen
      tf 02 cmd AddItem
      tf 03 evt ItemAdded
    `);
  });
  // ...
});
```

This is the primary visual correctness safeguard for a new diagram type and is required before merge.

---

### 🔴 [blocking] — Hardcoded colors break theming

`db.ts calculateEntityVisualProps()` uses hardcoded hex/named colors:
```typescript
case 'command': return { fill: '#bcd6fe', stroke: '#679ac3' };
case 'event':   return { fill: '#ffb778', stroke: '#c19a0f' };
// etc.
```

`styles.js` returns an empty string, so there are no CSS overrides either. Swimlane backgrounds (`rgb(250,250,250)`, `rgb(240,240,240)`) and the arrowhead fill (`#000`) in `renderer.ts` are also hardcoded.

Event modeling has semantic color conventions (blue=command, orange=event, green=readmodel, etc.) which is great — these should be exposed as **theme variables** that default to the canonical EM colors, rather than being burned in. This allows dark theme, custom themes, and accessibility overrides to function correctly.

See how `config.schema.yaml`'s `$defs/EventModelingDiagramConfig` could hold these colors with appropriate defaults, and `styles.js` could map them to CSS classes consumed by the renderer. The `journey` and `pie` diagrams are good references for this pattern.

---

### 🔴 [blocking] — Missing changeset

No `.changeset/*.md` file was generated. This is a new user-facing diagram feature and requires a `minor` changeset. Please run:
```bash
pnpm changeset
# select packages/mermaid, bump: minor
# message: feat: add Event Modeling diagram
```

---

### 🟡 [important] — Arrow marker ID not scoped to diagram

`renderer.ts` (near end of `draw()`) appends a marker with `id="arrowhead"`:
```typescript
diagram.append('defs').append('marker').attr('id', 'arrowhead')
```

This hardcoded ID will collide if multiple event modeling diagrams appear on the same page, and may conflict with other diagrams that use the same name. Use a per-render-scoped ID instead:
```typescript
.attr('id', `arrowhead-${id}`)
// ...
.attr('marker-end', `url(#arrowhead-${id})`)
```

---

### 🟡 [important] — `eventmodeling.spec.ts` fixture uses `>f` syntax — does not match grammar

`eventmodeling.spec.ts` fixture `fixtureMultipleRelations` uses:
```
tf 01 rmo CartScreen >f 02 >f 03 >f 04 >f 05
```

But the grammar (`event-modeling.langium`) defines source frame references as:
```
('->>' sourceFrames+=[EmFrame:EM_FI])*
```

`>f` is not valid grammar syntax — the correct operator is `->>`. It's worth checking whether this test actually passes and exercises what it claims. If Langium's error recovery silently ignores the unparseable `>f` tokens, `sourceFrames` would be empty, relations would be 0, and the assertion `expect(state.relations.length).toBe(4)` would fail. Please verify this test passes with `vitest run packages/mermaid/src/diagrams/eventmodeling` and correct the fixture if needed.

---

### 🟡 [important] — `populateCommonDb` is not called — accessibility and title won't work

`parser.ts` calls `db.setAst(ast)` but does not call `populateCommonDb(ast, db)`. The DB exports `setAccTitle`, `getAccTitle`, `setAccDescription`, `getAccDescription`, `setDiagramTitle`, and `getDiagramTitle` from `commonDb`, but without `populateCommonDb`, these will never be populated from the parsed AST. Diagram titles and accessibility attributes (`accTitle`, `accDescr`) won't work.

See how other Langium-based diagrams (e.g., `gitGraph`) call `populateCommonDb` in their `parser.ts`.

---

### 🟡 [important] — Config schema `rowHeight` description is copy-pasted from `packet`

`config.schema.yaml`:
```yaml
rowHeight:
  description: The height of each row in the packet diagram.  # ← should say "event modeling"
```

Also: `rowHeight` is defined in the config schema but doesn't appear to be used anywhere in the renderer or DB — is this intentional? If it's not used yet, consider removing it until it is, to avoid exposing unused config surface.

---

### 🟡 [important] — `setupGraphViewbox` called with `undefined` — `useMaxWidth` won't apply

`renderer.ts` end of `draw()`:
```typescript
setupGraphViewbox(undefined, diagram, 30, undefined);
```
The last argument is `useMaxWidth`, which comes from `DEFAULT_EVENTMODELING_CONFIG.useMaxWidth`. Passing `undefined` means `useMaxWidth` will never be applied, and the diagram won't respond correctly to the `useMaxWidth` config setting. Use:
```typescript
setupGraphViewbox(undefined, diagram, DEFAULT_EVENTMODELING_CONFIG.padding ?? 30, DEFAULT_EVENTMODELING_CONFIG.useMaxWidth);
```

---

### 🟢 [nit] — Stale `// @ts-ignore: JISON doesn't support types` comment in `diagram.ts`

`diagram.ts:1` has the comment `// @ts-ignore: JISON doesn't support types` but this diagram uses Langium, not JISON — and the `parser.ts` import doesn't need a `@ts-ignore`. Please remove.

### 🟢 [nit] — `getDirection()` returns a placeholder string

`db.ts`:
```typescript
export const getDirection = function () {
  return 'the direction'; // ← leftover placeholder
};
```

If this is unused, remove it. If it's required by the `DiagramDB` interface, return a sensible default or the actual direction.

### 🟢 [nit] — Large blocks of commented-out code

Several files have commented-out code that appears to be copy-pasted from gitGraph and never cleaned up: `db.ts` (`setOptions`/`getOptions` function bodies), `parser.ts` (the `populate` function), `renderer.ts` (multiple commented SVG rendering approaches). Please remove before merge to keep the initial implementation clean.

### 🟢 [nit] — `hasOwnProperty` called directly on object

`db.ts evolveFramePositioned`:
```typescript
if (state.swimlanes.hasOwnProperty(swimlaneProps.index)) {
```
Prefer `Object.prototype.hasOwnProperty.call(state.swimlanes, swimlaneProps.index)` or `swimlaneProps.index in state.swimlanes` to avoid prototype pollution concerns.

### 🟢 [nit] — `concept.md` and `pm.md` files in diagram folder

`packages/mermaid/src/diagrams/eventmodeling/concept.md` and `pm.md` appear to be working notes rather than documentation intended for users. These should either be removed or moved to an appropriate location (e.g., linked from `CONTRIBUTING.md` or kept outside the package source).

---

## Summary

The architecture and implementation quality are genuinely impressive for a first pass at a new diagram type. The two big things to resolve before this is ready to merge are the **XSS/sanitization issue** (user content must go through `sanitizeText()` before reaching `.html()`) and the **missing Cypress E2E tests**. The hardcoded colors and missing changeset are also required.

Happy to help clarify any of this — let's get it landed! 🚀

lgazo committed:
fix(eventmodeling): sanitize user input to prevent XSS attacks

User-provided text in entity names, inline data, and data blocks now
passes through sanitizeText() before being rendered via foreignObject.
This prevents malicious content from executing in the browser.

Fixes blocking issue from PR review #6440.

lgazo committed:
test(eventmodeling): add Cypress E2E visual regression tests

Add Cypress visual regression tests covering:
- State view pattern
- State change pattern
- Translation pattern
- Data block references
- Qualified names
- Multiple source relations

Also fix merge conflict marker in timings.json.

lgazo committed:
fix(eventmodeling): expose colors as theme variables and fix arrow marker ID

- Add theme variables for all entity colors (screen, processor, read model,
  command, event) with sensible defaults matching EM conventions
- Add theme variables for swimlane backgrounds and arrowhead color
- Scope arrow marker ID to diagram instance to avoid collisions
- Remove commented-out code from renderer

Theme variables added:
- emScreenFill, emScreenStroke
- emProcessorFill, emProcessorStroke
- emReadModelFill, emReadModelStroke
- emCommandFill, emCommandStroke
- emEventFill, emEventStroke
- emSwimlaneBackgroundOdd, emSwimlaneBackgroundStroke
- emArrowhead

Fixes blocking issue from PR review #6440.

lgazo committed:
fix(eventmodeling): update test fixture to use correct grammar syntax

The fixture fixtureMultipleRelations used invalid '>f' syntax.
Grammar defines multiple source frames using '->>' token.
This matches the Sequence diagram convention for consistency.

lgazo committed:
fix(eventmodeling): call populateCommonDb to enable accessibility and titles

Without populateCommonDb, diagram titles and accessibility attributes
(accTitle, accDescr) won't work. This ensures standard Mermaid
accessibility features are available for Event Modeling diagrams.

lgazo committed:
fix(eventmodeling): use proper config values for setupGraphViewbox

Pass config padding and useMaxWidth to setupGraphViewbox instead of
hardcoded values. Add padding config option to EventModelingDiagramConfig
schema with default of 30px.

lgazo committed:
chore(eventmodeling): clean up stale comments and fix code issues

- Remove stale @ts-ignore comment about JISON (diagram uses Langium)
- Fix getDirection() to return 'LR' instead of placeholder string
- Remove commented-out code in setOptions/getOptions
- Replace hasOwnProperty with 'in' operator
- Remove commented-out code in evolveFramePositioned

lgazo committed:
chore(eventmodeling): remove working notes from source

concept.md and pm.md are internal working notes, not user-facing
documentation. Remove them from the package source.

lgazo committed:
feat: add Event Modeling diagram

Adds Event Modeling diagram support, a method of describing systems
using an example of how information has changed within them over time.

Changeset for minor release.

lgazo committed:
test(eventmodeling): refactor unit tests to avoid DOM dependency

The previous tests called db.getState() which requires DOM for text
dimension calculation. Refactor tests to focus on parser capabilities
and add test cases for all entity types, data inline, data blocks,
reset frames, and multiple source frames.

lgazo committed:
test(examples): add eventmodeling example for test coverage

Add example diagram for eventmodeling to fix the test that requires
examples for all registered diagram types.

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
fix(eventmodeling): lint fixes

pbrolin47 commented:
Hi @lgazo 
I agree  the failing E2e, argos in this PR is not related to your changes. I rerun e2e (5) to get a clean run, but unfortunately we hit a screen-shot limitation on the license we are on. I raised the question to @ashishjain0512 and we addressing the problem

lgazo committed:
fix(eventmodeling): address PR retest feedback - themes, wrapLabel, relation stroke

- Add EM theme variables to all built-in themes (dark, default, forest, neutral)
  so dark mode and other themes render with appropriate colors instead of always
  falling back to light-theme defaults.
- Apply wrapLabel() to plain text before HTML assembly to prevent splitting
  inside HTML tags. Also remove redundant newline replacement in data block path
  since wrapLabel now handles line breaking on plain text.
- Read relation stroke color from themeVariables.emRelationStroke in renderer
  instead of using hardcoded '#000' from db.ts.
- Add emRelationStroke variable to all theme files including theme-base.

lgazo committed:
fix: type error in toHtml fixed

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo committed:
Merge branch 'develop' into feature/event-modeling-diagram

lgazo commented:
Hi @pbrolin47, all items should be fixed. I've merged it with the latest develop as well. There is one issue in e2e but that seems to be related to architecture diagram.

pbrolin47 reviewed (approved):
@lgazo. Final comments addressed. Thanks for your effort!

adymitruk commented:
Please follow the event->state view->processor->command->event pattern for any automation

adymitruk commented:
Also, state views can only attach to events from a single swimlane. This is because it's clear where the implementation for the state view will exist.

diegosasw commented:
> Also, state views can only attach to events from a single swimlane. This is because it's clear where the implementation for the state view will exist.

@adymitruk this raised a question for me.

I often model read models as projections across multiple streams, and I’ve been representing those as state views attached to events from multiple swimlanes.

How do you approach cases where a read model combines data from different aggregates? It makes me think my assumption that state views map directly to read models might not be quite right.

I can see the value of keeping a state view within a single swimlane if the goal is to show the information needed to make a decision (either automated or in a UI), independent of how it’s implemented. But since you mentioned implementation as the reason, I’d be interested in understanding that perspective a bit more.

adymitruk commented:
@diegosasw you first translate them into a single swimlane where the state view will live.
